cmd/oras/internal/display/progress/status.go

```diff
@@ -0,0 +1,207 @@
+	return &status{
+		prompt:     prompt,
+		descriptor: descriptor,
+		offset:     offset,
+	}
+}
+
+// StartTiming starts timing.
+func StartTiming() *status {
+	return &status{
+		offset:    -1,
+		startTime: time.Now(),
+	}
+}
+
+// EndTiming ends timing and set status to done.
+func EndTiming() *status {
+	return &status{
+		offset:  -1,
+		endTime: time.Now(),
+	}
+}
+
+func (s *status) isZero() bool {
+	return s.offset < 0 && s.startTime.IsZero() && s.endTime.IsZero()
+}
+
+// String returns human-readable TTY strings of the status.
+func (s *status) String(width int) (string, string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.isZero() {
+		return zeroStatus, zeroDigest
+	}
+	// todo: doesn't support multiline prompt
+	total := uint64(s.descriptor.Size)
+	var percent float64
+	if s.offset >= 0 {
+		percent = float64(s.offset) / float64(total)
+	}
+
+	name := s.descriptor.Annotations["org.opencontainers.image.title"]
+	if name == "" {
+		name = s.descriptor.MediaType
+	}
+
+	// format:  [left--------------------------------------------][margin][right---------------------------------]
+	//          mark(1) bar(22) speed(8) action(<=11) name(<=126)        size_per_size(<=13) percent(8) time(>=6)
+	//           └─ digest(72)
+	var offset string
+	switch percent {
+	case 1: // 100%, show exact size
+		offset = fmt.Sprint(s.total.size)
+	default: // 0% ~ 99%, show 2-digit precision
+		offset = fmt.Sprintf("%.2f", RoundTo(s.total.size*percent))
+	}
+	right := fmt.Sprintf(" %s/%v %s %6.2f%% %6s", offset, s.total.size, s.total.unit, percent*100, s.durationString())
+	lenRight := utf8.RuneCountInString(right)
+
+	var left string
+	lenLeft := 0
+	if !s.done {
+		lenBar := int(percent * barLength)
+		bar := fmt.Sprintf("[%s%s]", aec.Inverse.Apply(strings.Repeat(" ", lenBar)), strings.Repeat(".", barLength-lenBar))
+		speed := s.calculateSpeed()
+		speedStr := fmt.Sprintf("%v %2s/s", speed.size, speed.unit)
+		left = fmt.Sprintf("%c %s(%*s) %s %s", s.mark.symbol(), bar, speedLength, speedStr, s.prompt, name)
+		// bar + wrapper(2) + space(1) + speed + wrapper(2) = len(bar) + len(speed) + 5
+		lenLeft = barLength + speedLength + 5
+	} else {
+		left = fmt.Sprintf("√ %s %s", s.prompt, name)
+	}
+	// mark(1) + space(1) + prompt + space(1) + name = len(prompt) + len(name) + 3
+	lenLeft += utf8.RuneCountInString(s.prompt) + utf8.RuneCountInString(name) + 3
+
+	lenMargin := width - lenLeft - lenRight
+	if lenMargin < 0 {
+		// hide partial name with one space left
+		left = left[:len(left)+lenMargin-1] + "."
+		lenMargin = 0
+	}
+	return fmt.Sprintf("%s%s%s", left, strings.Repeat(" ", lenMargin), right), fmt.Sprintf("  └─ %s", s.descriptor.Digest.String())
+}
+
+// calculateSpeed calculates the speed of the progress and update last status.
+// caller must hold the lock.
+func (s *status) calculateSpeed() bytes {
+	now := time.Now()
+	secondsTaken := now.Sub(s.lastRenderTime).Seconds()
```

Comment:
What will happen when `s.lastRenderTime` is zero? Should we use `startTime` instead?

Reply:
@qweeah This still need to be addressed.